private/buf/bufsync/bufsyncapi/sync_handler.go

```diff
@@ -0,0 +1,526 @@
+
+	moduleIdentityToRepositoryIDCache  map[string]string
+	moduleIdentityToDefaultBranchCache map[string]string
+}
+
+func newSyncHandler(
+	logger *zap.Logger,
+	container appflag.Container,
+	repo git.Repository,
+	createWithVisibility string,
+	syncServiceClientFactory SyncServiceClientFactory,
+	referenceServiceClientFactory ReferenceServiceClientFactory,
+	repositoryServiceClientFactory RepositoryServiceClientFactory,
+	repositoryBranchServiceClientFactory RepositoryBranchServiceClientFactory,
+	repositoryTagServiceClientFactory RepositoryTagServiceClientFactory,
+	repositoryCommitServiceClientFactory RepositoryCommitServiceClientFactory,
+) bufsync.Handler {
+	return &syncHandler{
+		logger:                               logger,
+		container:                            container,
+		repo:                                 repo,
+		createWithVisibility:                 createWithVisibility,
+		moduleIdentityToRepositoryIDCache:    make(map[string]string),
+		moduleIdentityToDefaultBranchCache:   make(map[string]string),
+		syncServiceClientFactory:             syncServiceClientFactory,
+		referenceServiceClientFactory:        referenceServiceClientFactory,
+		repositoryServiceClientFactory:       repositoryServiceClientFactory,
+		repositoryBranchServiceClientFactory: repositoryBranchServiceClientFactory,
+		repositoryTagServiceClientFactory:    repositoryTagServiceClientFactory,
+		repositoryCommitServiceClientFactory: repositoryCommitServiceClientFactory,
+	}
+}
+
+func (h *syncHandler) ResolveSyncPoint(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branchName string,
+) (git.Hash, error) {
+	service := h.syncServiceClientFactory(moduleIdentity.Remote())
+	syncPoint, err := service.GetGitSyncPoint(ctx, connect.NewRequest(&registryv1alpha1.GetGitSyncPointRequest{
+		Owner:      moduleIdentity.Owner(),
+		Repository: moduleIdentity.Repository(),
+		Branch:     branchName,
+	}))
+	if err != nil {
+		if connect.CodeOf(err) == connect.CodeNotFound {
+			// No syncpoint
+			return nil, nil
+		}
+		return nil, fmt.Errorf("get git sync point: %w", err)
+	}
+	hash, err := git.NewHashFromHex(syncPoint.Msg.GetSyncPoint().GitCommitHash)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"invalid sync point from BSR %q: %w",
+			syncPoint.Msg.GetSyncPoint().GetGitCommitHash(),
+			err,
+		)
+	}
+	return hash, nil
+}
+
+func (h *syncHandler) IsGitCommitSynced(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	hash git.Hash,
+) (bool, error) {
+	service := h.referenceServiceClientFactory(moduleIdentity.Remote())
+	res, err := service.GetReferenceByName(ctx, connect.NewRequest(&registryv1alpha1.GetReferenceByNameRequest{
+		Owner:          moduleIdentity.Owner(),
+		RepositoryName: moduleIdentity.Repository(),
+		Name:           hash.Hex(),
+	}))
+	if err != nil {
+		if connect.CodeOf(err) == connect.CodeNotFound {
+			// Repo is not created.
```

Comment:
Or reference not found, I guess?

Reply:
No way to distinguish :(

Reply:
true :( for now updating the comment will do